bridge-ffi/src/lib.rs

```diff
@@ -0,0 +1,650 @@
+    unsafe {
+        if bytes.is_null() || (*bytes).disable_free {
+            return;
+        }
+    }
+    let bytes = bytes as *mut tmprl_bytes_t;
+    // Return vec back to core before dropping bytes
+    let vec = unsafe { Vec::from_raw_parts((*bytes).bytes as *mut u8, (*bytes).len, (*bytes).cap) };
+    // Set to null so the byte dropper doesn't try to free it
+    unsafe { (*bytes).bytes = std::ptr::null_mut() };
+    // Return only if core is non-null
+    if !core.is_null() {
+        let core = unsafe { &mut *core };
+        core.return_buf(vec);
+    }
+    unsafe {
+        Box::from_raw(bytes);
+    }
+}
+
+// Used for maintaining pointer to user data across threads
+struct UserDataHandle(*mut libc::c_void);
+unsafe impl Send for UserDataHandle {}
+unsafe impl Sync for UserDataHandle {}
+
+lazy_static::lazy_static! {
+    static ref DEFAULT_INIT_RESPONSE_BYTES: tmprl_bytes_t = {
+        tmprl_bytes_t::from_vec_disable_free(bridge::InitResponse::default().encode_to_vec())
+    };
+
+    static ref DEFAULT_SHUTDOWN_RESPONSE_BYTES: tmprl_bytes_t = {
+        tmprl_bytes_t::from_vec_disable_free(bridge::ShutdownResponse::default().encode_to_vec())
+    };
+
+    static ref DEFAULT_REGISTER_WORKER_RESPONSE_BYTES: tmprl_bytes_t = {
+        tmprl_bytes_t::from_vec_disable_free(bridge::RegisterWorkerResponse::default().encode_to_vec())
+    };
+}
+
+/// A runtime owned by Core. This must be passed to tmprl_runtime_free when no
+/// longer in use. This must not be freed until every call to every tmprl_core_t
+/// instance created with this runtime has been shutdown.
+pub struct tmprl_runtime_t {
+    tokio_runtime: std::sync::Arc<tokio::runtime::Runtime>,
+}
+
+/// Create a new runtime. The result is never null and must be freed via
+/// tmprl_runtime_free when no longer in use.
+#[no_mangle]
+pub extern "C" fn tmprl_runtime_new() -> *mut tmprl_runtime_t {
+    Box::into_raw(Box::new(tmprl_runtime_t {
+        // TODO(cretz): Options to configure thread pool?
+        tokio_runtime: std::sync::Arc::new(
+            tokio::runtime::Builder::new_multi_thread()
+                .enable_all()
+                .build()
+                .unwrap(),
+        ),
+    }))
+}
+
+/// Free a previously created runtime.
+#[no_mangle]
+pub extern "C" fn tmprl_runtime_free(runtime: *mut tmprl_runtime_t) {
+    if !runtime.is_null() {
+        unsafe {
+            Box::from_raw(runtime);
+        }
+    }
+}
+
+/// A core instance owned by Core. This must be passed to tmprl_core_shutdown
+/// when no longer in use which will free the resources.
+pub struct tmprl_core_t {
+    tokio_runtime: std::sync::Arc<tokio::runtime::Runtime>,
+    // TODO(cretz): Any concerns with the dynamic dispatch overhead here?
+    // TODO(cretz): Should we ask Rust SDK to give us back a struct instead?
+    // TODO(cretz): We could use generics, but impl traits don't cross the FFI
+    // boundary properly
+    core: std::sync::Arc<dyn temporal_sdk_core_api::Core>,
```

Comment:
In my latest PR the real struct may be used directly so we can do that, but I'm not concerned about dispatch overhead

Reply:
Cool. Documented that we accept the perf atm.